core/src/main/java/com/confighub/core/model/ConcurrentContextFilenameResponseCache.java

```diff
@@ -16,30 +14,29 @@ public static ConcurrentContextFilenameResponseCache getInstance()
         return instance;
     }
 
-    public Response put(Context context, String str, Response response)
+    public Response put(Repository repository, String context, String str, Response response)
     {
-        return put(new Pair<>(context, str), response);
+        synchronized (this)
```

Comment:
i'm concerned we lose all performance by locking here.  We should be using Concurrent dictionaries with locks around individual keys instead of locks around the whole collection.

Reply:
The majority of our lookups don't iterate the entire collection, so we only should have to lock the whole thing if we are adding something to the dictionary that didn't exist.  I assume the Java equivalent of C# ConcurrentDictionary works this way but i'm not sure.

Reply:
We are using a 2-level cache scheme.  If it was one level, we would just use ConcurrentHashMap, but this one use the repo name first and then the K/V pair after that.  The only changes I see here now is to remove the subclass' synchronized(this) and change to double-checked-locking in the base class to improve this.  I'll make those changes.